src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+        private BoundExpression MissingDeconstruct(BoundExpression receiver, ExpressionSyntax syntax, int numParameters, DiagnosticBag diagnostics, out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders, BoundNode childNode)
         {
             Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, receiver.Syntax, receiver.Type, numParameters);
             outPlaceholders = default(ImmutableArray<BoundDeconstructValuePlaceholder>);
 
             return BadExpression(syntax, childNode);
         }
+
+        public BoundStatement BindDeconstructionDeclarationStatementParts(LocalDeclarationStatementSyntax node, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Declaration.Deconstruction != null);
+
+            ArrayBuilder<DeconstructionVariable> variables = BindDeconstructionDeclarationVariables(node.Declaration, null, diagnostics);
+
+            var result = new BoundExpressionStatement(node, BindDeconstructionAssignment(node.Declaration.Deconstruction.Value, node.Declaration.Deconstruction.Value, variables, diagnostics));
+            FreeDeconstructionVariables(variables);
+
+            return result;
+        }
+
+        /// <summary>
+        /// The caller is responsible for releasing the nested ArrayBuilders.
+        /// </summary>
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax varSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            var variables = node.Deconstruction.Variables;
+
+            if (varSyntax == null)
+            {
+                varSyntax = node.Type;
+            }
+
+            // There are four cases for VariableDeclaration:
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. Thi could represent `(int x, ...)`
+
+            var localsBuilder = ArrayBuilder<BoundLocalDeclaration>.GetInstance(variables.Count);
+            var variablesBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                if (variable.Deconstruction == null)
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariable(variable, varSyntax, localsBuilder, diagnostics)));
+                }
+                else
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, varSyntax, diagnostics)));
+                }
```

Comment:
Consider moving condition inside `variablesBuilder.Add(...)`, or move `variablesBuilder.Add()` outside of `if`.